File: android/app/src/main/java/com/whimsyemu/drawingboard/SpeechBridge.java
package com.whimsyemu.drawingboard;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.speech.RecognitionListener;
import android.speech.RecognizerIntent;
import android.speech.SpeechRecognizer;
import android.webkit.JavascriptInterface;
import android.webkit.WebView;

import org.json.JSONObject;

import java.util.ArrayList;

/** Backs the page's "click to speak" button with Android's speech recogniser.
 *  Keeps listening across pauses until the page says stop. */
public class SpeechBridge {

    static final int REQUEST_AUDIO = 7;

    /** Injected into the page before its own scripts run (see MainActivity). */
    static final String POLYFILL =
        "(function(){ if (window.SpeechRecognition || window.webkitSpeechRecognition) return;" +
        " var active = null;" +
        " function SR(){ this.continuous=false; this.interimResults=false; this.lang='en-US'; this.onresult=null; this.onerror=null; this.onend=null; this.onstart=null; }" +
        " SR.prototype.start = function(){ active = this; try { AndroidSpeech.start(this.lang || 'en-US'); } catch(e){} };" +
        " SR.prototype.stop  = function(){ try { AndroidSpeech.stop(); } catch(e){} };" +
        " SR.prototype.abort = SR.prototype.stop;" +
        " window.__speechEvent = function(type, payload){ var r = active; if (!r) return; payload = payload || {};" +
        "   if (type === 'start')  { if (r.onstart) r.onstart({}); return; }" +
        "   if (type === 'result') { var alt = {transcript: String(payload.transcript || ''), confidence: 1}; var res = [alt]; res.isFinal = !!payload.isFinal; if (r.onresult) r.onresult({resultIndex: 0, results: [res]}); return; }" +
        "   if (type === 'error')  { if (r.onerror) r.onerror({error: payload.error || 'aborted'}); return; }" +
        "   if (type === 'end')    { if (active === r) active = null; if (r.onend) r.onend({}); } };" +
        " window.SpeechRecognition = SR; window.webkitSpeechRecognition = SR; })();";

    private final Activity activity;
    private final WebView web;
    private final Handler main = new Handler(Looper.getMainLooper());
    private SpeechRecognizer recognizer;
    private boolean listening = false;
    private boolean ended = true;
    private String lang = "en-US";

    SpeechBridge(Activity activity, WebView web) { this.activity = activity; this.web = web; }

    @JavascriptInterface
    public void start(String lang) { this.lang = lang == null ? "en-US" : lang; main.post(this::begin); }

    @JavascriptInterface
    public void stop() { main.post(() -> { listening = false; if (recognizer != null) recognizer.stopListening(); main.postDelayed(() -> finish(null), 1500); }); }

    private void emit(String type, String transcript, boolean isFinal, String error) {
        try {
            JSONObject o = new JSONObject();
            if (transcript != null) { o.put("transcript", transcript); o.put("isFinal", isFinal); }
            if (error != null) o.put("error", error);
            String js = "window.__speechEvent && window.__speechEvent(" + JSONObject.quote(type) + ", " + o + ");";
            main.post(() -> web.evaluateJavascript(js, null));
        } catch (Exception ignored) {}
    }

    private void finish(String error) {
        if (ended) return;
        ended = true; listening = false;
        if (recognizer != null) { try { recognizer.cancel(); } catch (Exception ignored) {} }
        if (error != null) emit("error", null, false, error);
        emit("end", null, false, null);
    }

    void onPermissionResult(boolean granted) { if (granted) begin(); else { ended = false; finish("not-allowed"); } }

    private void begin() {
        if (activity.checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            activity.requestPermissions(new String[]{Manifest.permission.RECORD_AUDIO}, REQUEST_AUDIO);
            return;
        }
        if (!SpeechRecognizer.isRecognitionAvailable(activity)) { ended = false; finish("audio-capture"); return; }
        if (recognizer == null) {
            recognizer = SpeechRecognizer.createSpeechRecognizer(activity);
            recognizer.setRecognitionListener(new RecognitionListener() {
                public void onReadyForSpeech(Bundle b) {}
                public void onBeginningOfSpeech() {}
                public void onRmsChanged(float v) {}
                public void onBufferReceived(byte[] bytes) {}
                public void onEndOfSpeech() {}
                public void onPartialResults(Bundle b) {
                    ArrayList<String> r = b.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
                    if (r != null && !r.isEmpty() && !ended) emit("result", r.get(0), false, null);
                }
                public void onResults(Bundle b) {
                    ArrayList<String> r = b.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
                    if (r != null && !r.isEmpty() && !ended) emit("result", r.get(0) + " ", true, null);
                    if (listening) listenOnce(); else finish(null);
                }
                public void onError(int code) {
                    if (ended) return;
                    boolean soft = code == SpeechRecognizer.ERROR_NO_MATCH || code == SpeechRecognizer.ERROR_SPEECH_TIMEOUT;
                    if (listening && soft) { listenOnce(); return; }
                    finish(code == SpeechRecognizer.ERROR_INSUFFICIENT_PERMISSIONS ? "not-allowed" : (soft ? null : "audio-capture"));
                }
                public void onEvent(int i, Bundle b) {}
            });
        }
        ended = false; listening = true;
        emit("start", null, false, null);
        listenOnce();
    }

    private void listenOnce() {
        Intent i = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)
                .putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
                .putExtra(RecognizerIntent.EXTRA_LANGUAGE, lang)
                .putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, true)
                .putExtra(RecognizerIntent.EXTRA_PREFER_OFFLINE, true);
        try { recognizer.startListening(i); } catch (Exception e) { finish("audio-capture"); }
    }

    void destroy() { if (recognizer != null) { recognizer.destroy(); recognizer = null; } }
}

File: ios/DrawingBoard/SpeechBridge.swift
import Foundation
import WebKit
import Speech
import AVFoundation

/// Gives the page a working `SpeechRecognition` so its "click to speak" button
/// works inside the app. The page thinks it is talking to the browser's speech
/// engine; we hand the audio to Apple's on-device recogniser (on-device whenever
/// the device supports it) and hand the words back.
final class SpeechBridge: NSObject, WKScriptMessageHandler {

    static let handlerName = "speech"

    /// Injected at document start, before the page checks whether speech exists.
    static let polyfill = """
    (function(){
      if (window.SpeechRecognition || window.webkitSpeechRecognition) return;
      var active = null;
      function post(msg){ try { window.webkit.messageHandlers.speech.postMessage(msg); } catch (e) {} }
      function SR(){ this.continuous = false; this.interimResults = false; this.lang = 'en-US';
                     this.onresult = null; this.onerror = null; this.onend = null; this.onstart = null; }
      SR.prototype.start = function(){ active = this; post({op:'start', lang: this.lang || 'en-US'}); };
      SR.prototype.stop  = function(){ post({op:'stop'}); };
      SR.prototype.abort = SR.prototype.stop;
      window.__speechEvent = function(type, payload){
        var r = active; if (!r) return; payload = payload || {};
        if (type === 'start')  { if (r.onstart) r.onstart({}); return; }
        if (type === 'result') { var alt = {transcript: String(payload.transcript || ''), confidence: 1};
                                 var res = [alt]; res.isFinal = !!payload.isFinal;
                                 if (r.onresult) r.onresult({resultIndex: 0, results: [res]}); return; }
        if (type === 'error')  { if (r.onerror) r.onerror({error: payload.error || 'aborted'}); return; }
        if (type === 'end')    { if (active === r) active = null; if (r.onend) r.onend({}); }
      };
      window.SpeechRecognition = SR; window.webkitSpeechRecognition = SR;
    })();
    """

    private weak var webView: WKWebView?
    private let audioEngine = AVAudioEngine()
    private var recognizer: SFSpeechRecognizer?
    private var request: SFSpeechAudioBufferRecognitionRequest?
    private var task: SFSpeechRecognitionTask?
    private var lastText = ""
    private var ended = true

    init(webView: WKWebView) {
        self.webView = webView
        super.init()
    }

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        guard let body = message.body as? [String: Any], let op = body["op"] as? String else { return }
        if op == "start" { start(lang: (body["lang"] as? String) ?? "en-US") } else { stop() }
    }

    private func emit(_ type: String, _ payload: [String: Any] = [:]) {
        guard let webView, let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8) else { return }
        DispatchQueue.main.async {
            webView.evaluateJavaScript("window.__speechEvent && window.__speechEvent(\"\(type)\", \(json));", completionHandler: nil)
        }
    }

    private func finish(withError error: String? = nil) {
        guard !ended else { return }
        ended = true
        teardown()
        if let error { emit("error", ["error": error]) }
        emit("end")
    }

    private func start(lang: String) {
        SFSpeechRecognizer.requestAuthorization { [weak self] status in
            guard let self else { return }
            guard status == .authorized else { self.ended = false; self.finish(withError: "not-allowed"); return }
            AVAudioSession.sharedInstance().requestRecordPermission { granted in
                guard granted else { self.ended = false; self.finish(withError: "not-allowed"); return }
                DispatchQueue.main.async { self.begin(lang: lang) }
            }
        }
    }

    private func begin(lang: String) {
        teardown()
        ended = false
        guard let rec = SFSpeechRecognizer(locale: Locale(identifier: lang)) ?? SFSpeechRecognizer(), rec.isAvailable else {
            finish(withError: "audio-capture"); return
        }
        recognizer = rec
        do {
            let session = AVAudioSession.sharedInstance()
            try session.setCategory(.record, mode: .measurement, options: .duckOthers)
            try session.setActive(true, options: .notifyOthersOnDeactivation)
        } catch { finish(withError: "audio-capture"); return }

        let req = SFSpeechAudioBufferRecognitionRequest()
        req.shouldReportPartialResults = true
        if rec.supportsOnDeviceRecognition { req.requiresOnDeviceRecognition = true }
        request = req

        let input = audioEngine.inputNode
        let format = input.outputFormat(forBus: 0)
        input.removeTap(onBus: 0)
        input.installTap(onBus: 0, bufferSize: 1024, format: format) { buffer, _ in req.append(buffer) }
        audioEngine.prepare()
        do { try audioEngine.start() } catch { finish(withError: "audio-capture"); return }

        lastText = ""
        task = rec.recognitionTask(with: req) { [weak self] result, error in
            guard let self, !self.ended else { return }
            if let result {
                self.lastText = result.bestTranscription.formattedString
                self.emit("result", ["transcript": self.lastText, "isFinal": result.isFinal])
                if result.isFinal { self.finish() }
            }
            if error != nil { self.finish() }
        }
        emit("start")
    }

    private func stop() {
        guard !ended else { return }
        request?.endAudio()
        if audioEngine.isRunning { audioEngine.stop() }
        audioEngine.inputNode.removeTap(onBus: 0)
        // If the recogniser doesn't deliver its final word promptly, close out with what we have.
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) { [weak self] in
            guard let self, !self.ended else { return }
            self.emit("result", ["transcript": self.lastText, "isFinal": true])
            self.finish()
        }
    }

    private func teardown() {
        if audioEngine.isRunning { audioEngine.stop() }
        audioEngine.inputNode.removeTap(onBus: 0)
        task?.cancel(); task = nil; request = nil
        try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
    }
}
